worker/firewaller/firewaller_test.go

```diff
@@ -6,128 +6,167 @@ package firewaller_test
-	"github.com/juju/juju/api/controller/remoterelations"
 	apitesting "github.com/juju/juju/api/testing"
-	"github.com/juju/juju/core/crossmodel"
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/core/life"
 	"github.com/juju/juju/core/network"
 	"github.com/juju/juju/core/network/firewall"
-	"github.com/juju/juju/core/status"
-	"github.com/juju/juju/environs"
+	"github.com/juju/juju/core/relation"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/watchertest"
 	"github.com/juju/juju/environs/config"
 	"github.com/juju/juju/environs/context"
 	"github.com/juju/juju/environs/instances"
-	"github.com/juju/juju/environs/models"
-	jujutesting "github.com/juju/juju/juju/testing"
-	"github.com/juju/juju/provider/dummy"
 	"github.com/juju/juju/rpc/params"
-	"github.com/juju/juju/state"
 	coretesting "github.com/juju/juju/testing"
-	"github.com/juju/juju/testing/factory"
 	"github.com/juju/juju/worker/firewaller"
+	"github.com/juju/juju/worker/firewaller/mocks"
 )
 
 const allEndpoints = ""
 
 // firewallerBaseSuite implements common functionality for embedding
 // into each of the other per-mode suites.
 type firewallerBaseSuite struct {
-	jujutesting.JujuConnSuite
-	charm              *state.Charm
-	controllerMachine  *state.Machine
-	controllerPassword string
+	testing.IsolationSuite
 
-	st                   api.Connection
-	firewaller           *apifirewaller.Client
-	remoteRelations      *remoterelations.Client
-	crossmodelFirewaller *crossmodelrelations.Client
-	clock                clock.Clock
+	firewaller           *mocks.MockFirewallerAPI
+	remoteRelations      *mocks.MockRemoteRelationsAPI
+	crossmodelFirewaller *mocks.MockCrossModelFirewallerFacadeCloser
+	credentialsFacade    *mocks.MockCredentialAPI
+	envFirewaller        *mocks.MockEnvironFirewaller
+	envModelFirewaller   *mocks.MockEnvironModelFirewaller
+	envInstances         *mocks.MockEnvironInstances
 
-	callCtx           context.ProviderCallContext
-	credentialsFacade *credentialvalidator.Facade
+	machinesCh     chan []string
+	applicationsCh chan struct{}
+	openedPortsCh  chan []string
+	remoteRelCh    chan []string
+	subnetsCh      chan []string
+	modelFwRulesCh chan struct{}
+
+	clock clock.Clock
+
+	firewallerStarted bool
+	modelFlushed      chan bool
+	machineFlushed    chan string
+	watchingMachine   chan names.MachineTag
+
+	mode                string
+	withIpv6            bool
+	withModelFirewaller bool
+
+	modelIngressRules firewall.IngressRules
+	envModelPorts     firewall.IngressRules
+
+	nextMachineId int
+	nextUnitId    map[string]int
+
+	deadMachines  set.Strings
+	instancePorts map[string]firewall.IngressRules
+	envPorts      firewall.IngressRules
+
+	mu             sync.Mutex
+	unitPortRanges *unitPortRanges
 }
 
 func (s *firewallerBaseSuite) SetUpTest(c *gc.C) {
-	s.JujuConnSuite.SetUpTest(c)
+	s.IsolationSuite.SetUpTest(c)
 
-	s.callCtx = context.NewEmptyCloudCallContext()
-}
+	s.modelFlushed = make(chan bool, 5)
```

Comment:
@wallyworld 
These are subject to races, which are causing CI failures. See:
https://jenkins.juju.canonical.com/job/unit-tests-race-amd64/1561/testReport/github/com_juju_juju_worker_firewaller/TestAll/